File: ports/php/src/Security/Boundary/AdmissibilityFormDrift.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Boundary;

use UnicodePhp\Security\Identity\Ucd;

// AdmissibilityFormDrift — cross-layer identifier-admissibility x form drift
// (boundary-layer detector).
//
// Byte-faithful transliteration of the verified Rust reference implementation,
// itself a mirror of the Lean specification for this family.
//
// Fires on inputs whose UTS #39 whole-string admissibility verdict differs
// between the input and its NFKC form. This is the string-level complement of
// IdentifierFormDrift (which scans Identifier_Status against the per-codepoint
// NFKD head): here the whole-string admissibility predicate is evaluated twice —
// once on the input, once on its NFKC form. The two are not redundant. In
// particular, a sequence of decomposed Hangul jamos passes the per-codepoint
// scan cleanly (each jamo has identity NFKD and Restricted status on both sides)
// but fires here: the jamo sequence is rejected by isAllowedIdentifier, while
// its NFKC composition into a precomposed Hangul syllable is accepted.
//
// It reuses the port's own UTS #39 admissibility predicate
// (Ucd::isAllowedIdentifier = UAX #31 default identifier AND every codepoint
// Allowed) and NFKC pipeline (Ucd::toNfkc), never a host normalization or
// identifier library.
//
// Sub-threat (direction-agnostic):
//   AdmissibilityFormDrift — isAllowedIdentifier(input) differs from
//   isAllowedIdentifier(toNfkc(input)). The pair of booleans is carried so the
//   verdict records which direction the drift goes; no position is reported
//   because the predicate is whole-string.

/**
 * Sub-threat carrier for AdmissibilityFormDrift. There is a single variant,
 * AdmissibilityFormDrift, holding the two whole-string admissibility booleans,
 * mirroring the rust `SubThreat` enum.
 */
final class AdmissibilityFormDriftSubThreat
{
    private function __construct(
        public readonly string $kind,
        public readonly bool $inputAdmissible,
        public readonly bool $nfkcAdmissible,
    ) {
    }

    public static function admissibilityFormDrift(bool $inputAdmissible, bool $nfkcAdmissible): AdmissibilityFormDriftSubThreat
    {
        return new AdmissibilityFormDriftSubThreat('AdmissibilityFormDrift', $inputAdmissible, $nfkcAdmissible);
    }

    /** Fixture-row tag string for this sub-threat (matches `SubThreat.tag`). */
    public function tag(): string
    {
        return match ($this->kind) {
            'AdmissibilityFormDrift' => 'AdmissibilityFormDrift',
            default => throw new \RuntimeException("AdmissibilityFormDriftSubThreat: unknown kind '{$this->kind}'"),
        };
    }
}

/**
 * Top-level classification (Clear when the admissibility verdict agrees across
 * forms), or a Hazard carrying the fired sub-threat, the implicated positions
 * (always empty — the predicate is whole-string), and the (always-empty)
 * decoded-byte projection kept for shape parity with the Lean
 * `Classification.hazard`.
 */
final class AdmissibilityFormDriftClassification
{
    /**
     * @param list<int> $positions
     * @param list<int> $decoded
     */
    private function __construct(
        public readonly ?AdmissibilityFormDriftSubThreat $sub,
        public readonly array $positions,
        public readonly array $decoded,
    ) {
    }

    public static function clear(): AdmissibilityFormDriftClassification
    {
        return new AdmissibilityFormDriftClassification(null, [], []);
    }

    /**
     * @param list<int> $positions
     * @param list<int> $decoded
     */
    public static function hazard(AdmissibilityFormDriftSubThreat $sub, array $positions, array $decoded): AdmissibilityFormDriftClassification
    {
        return new AdmissibilityFormDriftClassification($sub, $positions, $decoded);
    }

    /** True iff the classification is Clear. */
    public function isClear(): bool
    {
        return $this->sub === null;
    }

    /** Human-facing tag for a hazard, or null when clear. */
    public function tag(): ?string
    {
        return $this->sub?->tag();
    }

    /** @return list<int> Implicated positions (always empty — the predicate is whole-string). */
    public function positions(): array
    {
        return $this->positions;
    }
}

/** The structured output of `detect` (mirrors the Lean/rust `Verdict`). */
final class AdmissibilityFormDriftVerdict
{
    /** @param list<int> $input */
    public function __construct(
        public readonly array $input,
        public readonly AdmissibilityFormDriftClassification $classify,
        public readonly bool $inputAdmissible,
        public readonly bool $nfkcAdmissible,
    ) {
    }
}

final class AdmissibilityFormDrift
{
    /**
     * The AdmissibilityFormDrift detection function. Evaluates the whole-string
     * UTS #39 admissibility predicate on the input and on its NFKC form; a
     * disagreement is the hazard.
     * @param list<int> $input
     */
    public static function detect(array $input): AdmissibilityFormDriftVerdict
    {
        $input = array_values($input);
        $nfkc = Ucd::toNfkc($input);
        $inOk = Ucd::isAllowedIdentifier($input);
        $nfkcOk = Ucd::isAllowedIdentifier($nfkc);

        if ($inOk === $nfkcOk) {
            $classification = AdmissibilityFormDriftClassification::clear();
        } else {
            $classification = AdmissibilityFormDriftClassification::hazard(
                AdmissibilityFormDriftSubThreat::admissibilityFormDrift($inOk, $nfkcOk),
                [],
                [],
            );
        }

        return new AdmissibilityFormDriftVerdict($input, $classification, $inOk, $nfkcOk);
    }
}

File: ports/rust/src/security/boundary/admissibility_form_drift.rs
//! AdmissibilityFormDrift — cross-layer identifier-admissibility × form drift
//! (boundary-layer detector).
//!
//! Byte-faithful transliteration of
//! `Unicode/Security/Boundary/AdmissibilityFormDrift.lean`.
//!
//! Fires on inputs whose UTS #39 whole-string `isAllowedIdentifier` verdict
//! differs between the input and its NFKC form. This is the string-level
//! complement of IdentifierFormDrift (which scans `Identifier_Status` against the
//! per-codepoint NFKD head): here the whole-string admissibility predicate is
//! evaluated twice — once on the input, once on `to_nfkc(input)`. The two are not
//! redundant. In particular, a sequence of decomposed Hangul jamos passes the
//! per-codepoint scan cleanly (each jamo has identity NFKD and Restricted status
//! on both sides) but fires here: the jamo sequence is rejected by
//! `isAllowedIdentifier`, while its NFKC composition into a precomposed Hangul
//! syllable is accepted.
//!
//! It reuses the port's own UTS #39 admissibility predicate
//! (`ucd::is_allowed_identifier` = UAX #31 default identifier ∧ every codepoint
//! Allowed) and NFKC pipeline (`ucd::to_nfkc`), never a host normalization or
//! identifier library.
//!
//! Sub-threat (direction-agnostic):
//!   `AdmissibilityFormDrift` — `is_allowed_identifier(input) !=
//!   is_allowed_identifier(to_nfkc(input))`. The pair of booleans is carried so
//!   the verdict records which direction the drift goes; no position is reported
//!   because the predicate is whole-string.

use crate::security::identity::ucd;

// ─────────────────────────────────────────────────────────────────────
// §1 Types
// ─────────────────────────────────────────────────────────────────────

/// Sub-threat enumeration for AdmissibilityFormDrift.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum SubThreat {
    /// The whole-string admissibility verdict differs between the input and its
    /// NFKC form.
    AdmissibilityFormDrift {
        /// `is_allowed_identifier(input)`.
        input_admissible: bool,
        /// `is_allowed_identifier(to_nfkc(input))`.
        nfkc_admissible: bool,
    },
}

impl SubThreat {
    /// Fixture-row tag string for this sub-threat (matches `SubThreat.tag`).
    pub fn tag(&self) -> &'static str {
        match self {
            SubThreat::AdmissibilityFormDrift {
                input_admissible: _,
                nfkc_admissible: _,
            } => "AdmissibilityFormDrift",
        }
    }
}

/// Top-level classification for AdmissibilityFormDrift.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Classification {
    /// The admissibility verdict agrees across forms.
    Clear,
    /// The admissibility verdict drifts across forms.
    Hazard {
        /// The sub-threat that fired.
        sub: SubThreat,
        /// The implicated positions (always empty; the predicate is whole-string).
        positions: Vec<usize>,
        /// The decoded-byte projection (always empty here; shape parity with Lean).
        decoded: Vec<u8>,
    },
}

impl Classification {
    /// True iff the classification is `Clear`.
    pub fn is_clear(&self) -> bool {
        match self {
            Classification::Clear => true,
            Classification::Hazard {
                sub: _,
                positions: _,
                decoded: _,
            } => false,
        }
    }

    /// Human-facing tag for a hazard, or `None` when clear.
    pub fn tag(&self) -> Option<&'static str> {
        match self {
            Classification::Clear => None,
            Classification::Hazard {
                sub,
                positions: _,
                decoded: _,
            } => Some(sub.tag()),
        }
    }

    /// Implicated positions (always empty — the predicate is whole-string).
    pub fn positions(&self) -> &[usize] {
        match self {
            Classification::Clear => &[],
            Classification::Hazard {
                sub: _,
                positions,
                decoded: _,
            } => positions,
        }
    }
}

/// The structured output of `detect` (mirrors the Lean `Verdict`).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Verdict {
    /// The scanned input codepoints.
    pub input: Vec<u32>,
    /// The classification verdict.
    pub classify: Classification,
    /// `is_allowed_identifier(input)`.
    pub input_admissible: bool,
    /// `is_allowed_identifier(to_nfkc(input))`.
    pub nfkc_admissible: bool,
}

// ─────────────────────────────────────────────────────────────────────
// §2 Top-level detection
// ─────────────────────────────────────────────────────────────────────

/// The AdmissibilityFormDrift detection function.
pub fn detect(input: &[u32]) -> Verdict {
    let nfkc = ucd::to_nfkc(input);
    let in_ok = ucd::is_allowed_identifier(input);
    let nfkc_ok = ucd::is_allowed_identifier(&nfkc);

    let classification = if in_ok == nfkc_ok {
        Classification::Clear
    } else {
        Classification::Hazard {
            sub: SubThreat::AdmissibilityFormDrift {
                input_admissible: in_ok,
                nfkc_admissible: nfkc_ok,
            },
            positions: Vec::new(),
            decoded: Vec::new(),
        }
    };

    Verdict {
        input: input.to_vec(),
        classify: classification,
        input_admissible: in_ok,
        nfkc_admissible: nfkc_ok,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::security::calculus::Family;
    use crate::security::policy::reason_code;

    // Ground truth: every `detect_*` theorem in
    // `Unicode/Security/Boundary/AdmissibilityFormDrift.lean`. Each maps to one
    // `#[test]`; the shared context-free fixture
    // (`fixtures/security/detectors/admissibility_form_drift.json`) carries the
    // same vectors.

    fn tag(input: &[u32]) -> Option<&'static str> {
        detect(input).classify.tag()
    }

    // `detect_empty_clear` — both admissibility calls return false, so they agree.
    #[test]
    fn detect_empty_clear() {
        assert!(detect(&[]).classify.is_clear());
    }

    // `detect_ascii_clear` — "admin"; admissible on both sides (NFKC is identity).
    #[test]
    fn detect_ascii_clear() {
        let v = detect(&[0x61, 0x64, 0x6D, 0x69, 0x6E]);
        assert!(v.classify.is_clear());
        assert!(v.input_admissible);
        assert!(v.nfkc_admissible);
    }

    // `detect_fi_ligature_drift` — ﬁ (U+FB01) is Restricted (inadmissible), but
    // NFKC decomposes it to "fi" (admissible). Drift fires.
    #[test]
    fn detect_fi_ligature_drift() {
        let v = detect(&[0xFB01]);
        assert_eq!(v.classify.tag(), Some("AdmissibilityFormDrift"));
        assert!(!v.input_admissible);
        assert!(v.nfkc_admissible);
    }

    // `detect_jamo_sequence_drift` — decomposed Hangul jamos [U+1112, U+1161,
    // U+11AB] are inadmissible, but NFKC composes them to U+D55C 한 (admissible).
    #[test]
    fn detect_jamo_sequence_drift() {
        assert_eq!(tag(&[0x1112, 0x1161, 0x11AB]), Some("AdmissibilityFormDrift"));
    }

    // The composed reason code for the sole sub-threat.
    #[test]
    fn reason_code_is_stable() {
        assert_eq!(
            reason_code(Family::AdmissibilityFormDrift, Some("AdmissibilityFormDrift")),
            "unicode.security.X.admissibility-form-drift.AdmissibilityFormDrift"
        );
    }
}
